client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoStatement.java

```diff
@@ -86,7 +87,7 @@ public void close()
     {
         TrinoConnection connection = this.connection.getAndSet(null);
         if (connection != null) {
-            connection.removeStatement(this);
+            onClose.accept(this);
```

Comment:
the `if (connection != null)` guarding this is now redundant

if you want to have it to make close idempotent, it would be better to make it explicit

```
TrinoConnection connection = this.connection.getAndSet(null);
if (connection == null) {
    // already closed
   return
}
```

Reply:
> the if (connection != null) guarding this is now redundant

No, It was needed because you `unregisterStatement` requires statement to be first registered.

> if you want to have it to make close idempotent, it would be better to make it explicit

That changes the preexisting logic, where `closeResultSet()` was always called. Will extract this to different commit.

Reply:
> No, It was needed because you `unregisterStatement` requires statement to be first registered.

i am not sure i understand this.
The connection-null-check doesn't tell you whether it was registered already or not.

However, note that if you consolidate initialization responsibility in `createStatement`  (https://github.com/trinodb/trino/pull/7819#discussion_r631832815)
there is no possibility that you call uregister on a partially initialized (not registered yet) statement